src/Text/LLVM/PP.hs

```diff
@@ -1279,6 +1302,30 @@ opt :: Bool -> Fmt Doc
 opt True  = id
 opt False = const empty
 
+-- | Print a ValMd' value as an Int64.  If the ValMd' is not representable as an
+-- Int64, simply print nothing, which is where it differs from ppValMd' which
+-- will print *any* metadata value.
+
+ppInt64ValMd' :: Fmt (ValMd' i)
+ppInt64ValMd' = \case
+  ValMdValue tv
+    | PrimType (Integer _) <- typedType tv
+    , ValInteger i <- typedValue tv
+      -> integer i  -- 64 bits is the largest Int, so no conversion needed
+  ValMdDebugInfo (DebugInfoGlobalVariable gv) ->
+    case digvVariable gv of
+      Nothing -> mempty
+      Just v -> ppInt64ValMd' v
+  ValMdDebugInfo (DebugInfoGlobalVariableExpression expr) ->
+    case digveExpression expr of
+      Nothing -> mempty
+      Just e -> ppInt64ValMd' e
+  ValMdDebugInfo (DebugInfoLocalVariable lv) ->
+    integer $ fromIntegral $ dilvArg lv  -- ??
+  ValMdRef _idx -> mempty -- no table here to look this up...
+  _ -> mempty  -- TODO: generate warning?
```

Comment:
This code is quite complicated, and (I think) needlessly so. Why not just `ppValMd'` instead of `ppInt64ValMd'`? This code seems to have aspirations of checking for invalid states, but I don't think the pretty-printing code is the right place to do this. (Besides, we don't check for invalid states in `count`/`lowerBound` on recent LLVM versions, so I don't really see the point in doing so here.)

Reply:
This code is invoked for specific earlier versions of LLVM where the corresponding field is stored in the bitcode as a direct signed value instead of a metadata reference; using `ppValMd'` will emit a metadata reference.  This is not checking for invalid states, but instead this is the projection of a signed integer value back out of a `ValMd`.

If the PP is not the right location to adjust the output to be specific to the LLVM version we are emitting, where would you recommend doing this?

Reply:
Thank you for pushing back here, since I now realize that the way I imagined that `DISubrange` should be pretty-printed was slightly off. In my mind, I had thought that pretty-printing as `Int64` value directly would result in the same as `Doc` as pretty-printing a `ValMdValue (Typed (PrimType (Integer 64)) (ValInteger x))` value (which is what these `Int64` values get turned into on old LLVM versions). Now that I look more closely, however, that isn't the case at all. In `Text.LLVM.PP`, we have:

```hs
ppValMd' :: Fmt i -> Fmt (ValMd' i)
ppValMd' pp m = case m of
  ...
  ValMdValue tv     -> ppTyped (ppValue' pp) tv
  ...

ppTyped :: Fmt a -> Fmt (Typed a)
ppTyped fmt ty = ppType (typedType ty) <+> fmt (typedValue ty)

ppValue' :: Fmt i -> Fmt (Value' i)
ppValue' pp val = case val of
  ValInteger i       -> integer i
  ...
```

The key function here is `ppTyped`. This would cause, say, `ValMdValue (Typed (PrimType (Integer 64)) (ValInteger 27))` to be pretty-printed as `i64 27`, not `27`! This is definitely not what we want.

But looking even more closely, I think the same mistake also applies to the code is used on recent LLVM versions. Above in `ppDISubrange'`, we have:

```hs
                  let ppV = if llvmVer >= 7
                            then ppValMd' pp
                            else ppInt64ValMd'
                  in (("count:" <+>) . ppV) <$> disrCount sr
```

If I'm reading this correctly, then this code would have the potential to mistakenly print something like `count: i64 27` rather than `count: 27` on LLVM 7 and later. (I haven't tested this hypothesis, but it looks very likely.) Looking at the [LLVM repo's test cases](https://github.com/search?q=repo%3Allvm%2Fllvm-project+disubrange+language%3ALLVM&type=code&l=LLVM) (and the [code for pretty-printing `DISubrange`s](https://github.com/llvm/llvm-project/blob/431969ede13b0974041e948393d94bca1720c5d7/llvm/lib/IR/AsmWriter.cpp#L1888-L1927)), I don't think this was the intention.

As such, I've reluctantly come to the conclusion that we do need something like `ppInt64ValMd'`, and not just for old LLVM versions, but for _all_ LLVM versions. It's a bit grotesque, but then again, the [upstream LLVM code for pretty-printing `DISubrange`s](https://github.com/llvm/llvm-project/blob/431969ede13b0974041e948393d94bca1720c5d7/llvm/lib/IR/AsmWriter.cpp#L1888-L1927) is also grotesque, so it's hard to avoid.

Reply:
OK, I adjusted for this, and it is passing all current tests (llvm-pretty and llvm-pretty-bc-parser).  We may need to revisit this as our tests and/or LLVM version ranges broaden, but I think this is a good starting point.  Thanks for your investigations on this as well: I think we are in a better spot for it.